openedx/core/djangoapps/content_libraries/views.py

```diff
@@ -689,3 +696,130 @@ def delete(self, request, usage_key_str, file_path):
         except ValueError:
             raise ValidationError("Invalid file path")  # lint-amnesty, pylint: disable=raise-missing-from
         return Response(status=status.HTTP_204_NO_CONTENT)
+
+
+@view_auth_classes()
+class LibraryImportTaskViewSet(ViewSet):
+    """
+    Import blocks from Courseware through modulestore.
+    """
+
+    @convert_exceptions
+    def list(self, request, lib_key_str):
+        """
+        List all import tasks for this library.
+        """
+        library_key = LibraryLocatorV2.from_string(lib_key_str)
+        api.require_permission_for_library_key(
+            library_key,
+            request.user,
+            permissions.CAN_VIEW_THIS_CONTENT_LIBRARY
+        )
+        queryset = api.ContentLibrary.objects.get_by_key(library_key).import_tasks
+        result = ContentLibraryBlockImportTaskSerializer(queryset, many=True).data
+        paginator = LibraryApiPagination()
+        return paginator.get_paginated_response(
+            paginator.paginate_queryset(result, request)
+        )
+
+    @convert_exceptions
+    def create(self, request, lib_key_str):
+        """
+        Create and queue an import tasks for this library.
+        """
+
+        library_key = LibraryLocatorV2.from_string(lib_key_str)
+        api.require_permission_for_library_key(
+            library_key,
+            request.user,
+            permissions.CAN_EDIT_THIS_CONTENT_LIBRARY,
+        )
+
+        serializer = ContentLibraryBlockImportTaskCreateSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        course_key = serializer.validated_data['course_key']
+
+        import_task = api.import_blocks_create_task(library_key, course_key)
+        return Response(ContentLibraryBlockImportTaskSerializer(import_task).data)
+
+    @convert_exceptions
+    def retrieve(self, request, lib_key_str, pk=None):
+        """
+        Retrieve a import task for inspection.
+        """
+
+        library_key = LibraryLocatorV2.from_string(lib_key_str)
+        api.require_permission_for_library_key(
+            library_key,
+            request.user,
+            permissions.CAN_VIEW_THIS_CONTENT_LIBRARY,
+        )
+
+        import_task = api.ContentLibraryBlockImportTask.objects.get(pk=pk)
+        return Response(ContentLibraryBlockImportTaskSerializer(import_task).data)
+
+
+@view_auth_classes()
+class LibraryImportableCourseView(APIView):
+    """
+    List libraries that can be imported by the given user.
```

Comment:
This docstring and the `query_parameter` docs below are wrong, because this API is only about courses, not libraries.

In fact, I don't really think this belongs in the libraries REST API area because it has nothing to do with libraries. It only lists courses that the user has Studio access to.

I believe that our current LabXchange import code makes one call to the existing `/api/courses/v1/courses/` API to get all courses the user can read, then makes one call to `/api/enrollment/v1/` to get all courses that the user has Instructor access to, and then combines that data.

So I don't think that a new API is needed, but if you want to make it easier for API clients, can we just an add optional query parameter to the existing `/api/courses/v1/courses/` endpoint to allow filtering by role?

Reply:
> So I don't think that a new API is needed, but if you want to make it easier for API clients, can we just add an optional query parameter to the existing /api/courses/v1/courses/ endpoint to allow filtering by role?

I totally agree here. But it is not immediately trivial to include the filter to `/api/courses/v1/courses`, and it would take some additional effort to get there. My suggestion is, let's go with the two API calls while we plan the work on expanding that API to accommodate this use case, since this is obviously the long-term solution.

We would have to update our MFE patch @gabor-boros. Any blockers against this approach?

Reply:
@jvdm We thought about this approach with @arbrandes, though in the middle of the implementation of the frontend I proposed to have this endpoint to simplify API client calls and I found a blocker that time.

Since then, the frontend implementation has been changed, and this approach could be solved on the frontend if needed, however, I'm a bit afraid that the API calls would be pretty slow in that case.

Reply:
@gabor-boros Why would those two API calls be significantly slower than this one? (I'm assuming you can make the two calls in parallel btw)

Reply:
I cannot remember exactly how the courses API response fields are prepared on the backend. As far as I remember -- I may remember incorrectly -- the data returned by that API collected from several places that slows the processing of 50 course responses compared to this API.

Though I'd like to highlight again that I thought about the solution you proposed in the first place.

Reply:
I think we can move on from [this comment from the reviewer](https://github.com/edx/edx-platform/pull/27715#issuecomment-895469980) with the current API and capture this as a follow-up. I can capture the words to add a comment to the patch to explain why we are pushing this. Let me know if someone disagrees.

@gabor-boros:

> the data returned by that API collected from several places that slows the processing of 50 course responses compared to this API.

How slow was it? Would it impact UX that much. I think the course import user flow is not exercised frequently by the same user, and maybe we are prematurely optimizing for performance but harnessing the separation of concerns?

Reply:
@jvdm I have no measurements from real usage.

Reply:
@jvdm, @gabor-boros, have we considered pagination?  Wouldn't splitting the API call make paginating in the frontend unreliable?

Reply:
@arbrandes @gabor-boros We already use the existing two APIs from the LabXchange app for this _exact_ same use case, and I believe it works fine. To handle the case of staff users who have access to hundreds of courses, the backend caches (per-user) the full list of courses that a user can see. The frontend can then get that data from the backend, paginated however it wants.

Code (not open source) is [here](https://gitlab.com/opencraft/client/LabXchange/labxchange-dev/-/blob/616c1543b96c4cc331f541eeab48654f98fcd6c6/backend/labxchange/service_clients/edxapp/client.py#L665-724).

In the new use case here, there is no separate backend to do caching but the frontend is presumably a SPA so it can just retrieve and cache the entire list of courses, then cross-reference it with the list of roles and display it paginated. In fact it's arguably better to do that on the frontend so you can display a progress bar / loading message while it loads.

Reply:
@bradenmacdonald, > it can just retrieve and cache the entire list of courses, then cross-reference it with the list of roles and display it paginated.

I am suspicious of doing frontend pagination based on two or more disparate, and potentially large, data sources.  Sure, it's doable, but it is complicated, second-guesses DRF, and is generally worse as far as frontend performance goes.  Sure, if Open edX gives us no other solution that's what we'll have to do, but I think we should at least aim towards porting the functionality to the right namespace.

Reply:
Addendum: the user-facing side supports searching, which would further the frontend (and pagination) further.

Reply:
OK. My bottom line is that the endpoint in question is a type of course search/list API and it belongs in the same namespace as the other course search/list APIs, or perhaps even should be just a filter/parameter on one of those existing APIs. And because of that, I think it should at least be submitted as a separate PR so as not to hold up the rest of this course import functionality, and to make it easier for people who work on the course APIs to be aware of it and comment.